File: apps/web/services/resources/resources.ts
import { getAPIUrl } from '@services/config/config'
import {
  RequestBodyFormWithAuthHeader,
  RequestBodyWithAuthHeader,
  errorHandling,
  getResponseMetadata,
} from '@services/utils/ts/requests'

export type ResourceType =
  | 'assessment'
  | 'video'
  | 'article'
  | 'tool'
  | 'guide'
  | 'course'
  | 'other'

export interface ResourceChannel {
  id: number
  org_id: number
  channel_uuid: string
  name: string
  description: string | null
  thumbnail_image: string | null
  public: boolean
  shared: boolean
  is_starred: boolean
  color: string | null
  owner_org_id?: number | null
  owner_org_uuid?: string | null
  owner_org_slug?: string | null
  owner_org_name?: string | null
  is_shared_from_other_org?: boolean
  creation_date: string
  update_date: string
  resource_count: number
  is_accessible: boolean
  usergroup_ids: number[]
}

export interface UserResourceChannel {
  id: number
  user_id: number
  org_id: number
  user_channel_uuid: string
  name: string
  description: string | null
  is_default: boolean
  icon: string | null
  color: string | null
  icon_color: string | null
  creation_date: string
  update_date: string
  resource_count: number
}

export interface SavedResourceState {
  id: number
  user_id: number
  resource_id: number
  notes: string | null
  outcome_text: string | null
  outcome_link: string | null
  outcome_file: string | null
  last_opened_at: string | null
  completed_at: string | null
  open_count: number
  creation_date: string
  update_date: string
}

export interface Resource {
  id: number
  org_id: number
  resource_uuid: string
  created_by_user_id: number | null
  title: string
  description: string | null
  resource_type: ResourceType
  provider_name: string | null
  provider_url: string | null
  external_url: string
  cover_image_url: string | null
  thumbnail_image: string | null
  owner_org_id?: number | null
  owner_org_uuid?: string | null
  owner_org_slug?: string | null
  owner_org_name?: string | null
  is_shared_from_other_org?: boolean
  estimated_time: number | null
  is_featured: boolean
  is_live: boolean
  access_mode: 'free' | 'paid' | 'restricted'
  creation_date: string
  update_date: string
  save_count: number
  comment_count: number
  average_rating: number | null
  rating_count: number
  tags: ResourceTag[]
  is_saved: boolean
  has_outcome: boolean
  channels: ResourceChannel[]
  user_channel_uuids: string[]
  user_state: SavedResourceState | null
}

export interface ResourceTag {
  id: number
  org_id: number
  tag_uuid: string
  name: string
  creation_date: string
  update_date: string
}

export interface ResourceComment {
  id: number
  resource_id: number
  author_id: number
  comment_uuid: string
  content: string
  rating: number | null
  is_locked: boolean
  creation_date: string
  update_date: string
  author?: {
    id: number
    user_uuid: string
    username: string
    first_name: string
    last_name: string
    avatar_image: string | null
  } | null
}

export type ResourceNoteBlockType = 'text' | 'link' | 'image' | 'file'

export interface ResourceNoteBlock {
  id: number
  user_id: number
  resource_id: number
  note_uuid: string
  block_type: ResourceNoteBlockType
  media_asset_uuid: string | null
  content: string | null
  url: string | null
  title: string | null
  description: string | null
  preview_image_url: string | null
  filename: string | null
  original_filename: string | null
  mime_type: string | null
  storage_directory: string | null
  sort_order: number
  creation_date: string
  update_date: string
}

export interface ResourceUrlPreview {
  title?: string | null
  description?: string | null
  og_image?: string | null
  favicon?: string | null
  og_type?: string | null
  og_url?: string | null
  url?: string | null
}

export interface GetResourcesParams {
  channel_uuid?: string
  user_channel_uuid?: string
  resource_type?: ResourceType
  resource_types?: string
  tags?: string
  provider?: string
  query?: string
  access?: 'free' | 'paid' | 'restricted' | string
  saved_only?: boolean
  completed_only?: boolean
  include_private?: boolean
  offset?: number
  limit?: number
}

export async function getResourceChannels(orgId: number, accessToken?: string, includePrivate = false) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}/channels?include_private=${includePrivate}`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<{ channels: ResourceChannel[]; user_channels: UserResourceChannel[] }>
}

export async function getResourceTags(orgId: number, accessToken?: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}/tags`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceTag[]>
}

export async function createResourceTag(orgId: number, data: { name: string }, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}/tags`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceTag>
}

export async function updateResourceTag(tagUuid: string, data: { name: string }, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/tags/${tagUuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceTag>
}

export async function deleteResourceTag(tagUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/tags/${tagUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function getResources(
  orgId: number,
  params: GetResourcesParams,
  accessToken?: string
) {
  const search = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '' && value !== false) search.set(key, String(value))
  })
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}?${search.toString()}`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<Resource[]>
}

export async function getResource(resourceUuid: string, accessToken?: string, includePrivate = false) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}?include_private=${includePrivate}`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<Resource>
}

export async function createResource(orgId: number, data: any, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/?org_id=${orgId}`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function updateResource(resourceUuid: string, data: any, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function deleteResource(resourceUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function uploadResourceThumbnail(resourceUuid: string, formData: FormData, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/thumbnail`,
    RequestBodyFormWithAuthHeader('PUT', formData, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function createResourceChannel(orgId: number, data: any, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/channels?org_id=${orgId}`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function updateResourceChannel(channelUuid: string, data: any, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/channels/${channelUuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function deleteResourceChannel(channelUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/channels/${channelUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function uploadResourceChannelThumbnail(channelUuid: string, formData: FormData, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/channels/${channelUuid}/thumbnail`,
    RequestBodyFormWithAuthHeader('PUT', formData, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function getChannelResources(
  channelUuid: string,
  accessToken?: string,
  includePrivate = false
) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/channels/${channelUuid}/resources?include_private=${includePrivate}`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<Resource[]>
}

export async function addResourceToChannel(
  channelUuid: string,
  resourceUuid: string,
  accessToken: string,
  sortOrder = 0
) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/channels/${channelUuid}/resources`,
    RequestBodyWithAuthHeader('POST', { resource_uuid: resourceUuid, sort_order: sortOrder }, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function removeResourceFromChannel(channelUuid: string, resourceUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/channels/${channelUuid}/resources/${resourceUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function createUserResourceChannel(orgId: number, data: any, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}/me/channels`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function updateUserResourceChannel(
  orgId: number,
  userChannelUuid: string,
  data: any,
  accessToken: string
) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}/me/channels/${userChannelUuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function deleteUserResourceChannel(orgId: number, userChannelUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}/me/channels/${userChannelUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function saveResource(resourceUuid: string, data: any, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/save`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function unsaveResource(resourceUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/save`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function uploadOutcomeFile(resourceUuid: string, formData: FormData, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/save/outcome-file`,
    RequestBodyFormWithAuthHeader('PUT', formData, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function getResourceNotes(resourceUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/notes`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceNoteBlock[]>
}

export async function createResourceNote(
  resourceUuid: string,
  data: {
    block_type: ResourceNoteBlockType
    media_asset_uuid?: string
    content?: string
    url?: string
    title?: string
    preview_image_url?: string
  },
  accessToken: string
) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/notes`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceNoteBlock>
}

export async function uploadResourceNote(resourceUuid: string, file: File, accessToken: string) {
  const formData = new FormData()
  formData.append('file', file)
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/notes/upload`,
    RequestBodyFormWithAuthHeader('POST', formData, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceNoteBlock>
}

export async function updateResourceNote(
  noteUuid: string,
  data: Partial<Pick<ResourceNoteBlock, 'content' | 'url' | 'title' | 'description' | 'preview_image_url' | 'sort_order'>>,
  accessToken: string
) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/notes/${noteUuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceNoteBlock>
}

export async function deleteResourceNote(noteUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/notes/${noteUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function getResourceComments(resourceUuid: string, accessToken?: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/comments`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceComment[]>
}

export async function createResourceComment(resourceUuid: string, data: { content: string }, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/comments`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceComment>
}

export async function updateResourceComment(commentUuid: string, data: { content: string }, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/comments/${commentUuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceComment>
}

export async function deleteResourceComment(commentUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/comments/${commentUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function getResourceReviews(resourceUuid: string, accessToken?: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/reviews`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceComment[]>
}

export async function createResourceReview(
  resourceUuid: string,
  data: { content: string; rating: number },
  accessToken: string
) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/${resourceUuid}/reviews`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceComment>
}

export async function updateResourceReview(
  commentUuid: string,
  data: { content?: string; rating?: number },
  accessToken: string
) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/reviews/${commentUuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceComment>
}

export async function deleteResourceReview(commentUuid: string, accessToken: string) {
  const result: any = await fetch(
    `${getAPIUrl()}resources/reviews/${commentUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function importResourcesCsv(
  orgId: number,
  formData: FormData,
  accessToken: string,
  channelUuid?: string
) {
  const search = new URLSearchParams()
  if (channelUuid) search.set('channel_uuid', channelUuid)
  const result: any = await fetch(
    `${getAPIUrl()}resources/org/${orgId}/import${search.toString() ? `?${search.toString()}` : ''}`,
    RequestBodyFormWithAuthHeader('POST', formData, null, accessToken)
  )
  return getResponseMetadata(result)
}

export async function getResourceUrlPreview(url: string, accessToken?: string) {
  const result: any = await fetch(
    `${getAPIUrl()}utils/link-preview?url=${encodeURIComponent(url)}`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result) as Promise<ResourceUrlPreview>
}

File: apps/api/src/db/resources.py
from enum import Enum

from sqlalchemy import Boolean, Column, ForeignKey, Integer, Text, UniqueConstraint
from sqlmodel import Field, SQLModel


class ResourceTypeEnum(str, Enum):
    assessment = "assessment"
    video = "video"
    article = "article"
    tool = "tool"
    guide = "guide"
    course = "course"
    other = "other"


class ResourceAccessModeEnum(str, Enum):
    free = "free"
    paid = "paid"
    restricted = "restricted"


class ResourceBase(SQLModel):
    title: str
    description: str | None = Field(default=None, sa_column=Column(Text))
    resource_type: ResourceTypeEnum = ResourceTypeEnum.other
    provider_name: str | None = None
    provider_url: str | None = None
    external_url: str
    cover_image_url: str | None = None
    thumbnail_image: str | None = None
    estimated_time: int | None = None
    is_featured: bool = False
    is_live: bool = False
    access_mode: ResourceAccessModeEnum = ResourceAccessModeEnum.free


class ResourceTagBase(SQLModel):
    name: str


class ResourceTag(ResourceTagBase, table=True):
    __tablename__ = "resourcetag"
    __table_args__ = (UniqueConstraint("org_id", "name", name="uq_resourcetag_org_name"),)

    id: int | None = Field(default=None, primary_key=True)
    org_id: int = Field(
        sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True)
    )
    tag_uuid: str = Field(index=True, unique=True)
    creation_date: str = ""
    update_date: str = ""


class ResourceTagCreate(ResourceTagBase):
    pass


class ResourceTagUpdate(SQLModel):
    name: str | None = None


class ResourceTagRead(ResourceTagBase):
    id: int
    org_id: int
    tag_uuid: str
    creation_date: str
    update_date: str


class ResourceTagLink(SQLModel, table=True):
    __tablename__ = "resourcetaglink"
    __table_args__ = (
        UniqueConstraint("resource_id", "tag_id", name="uq_resource_tag"),
    )

    id: int | None = Field(default=None, primary_key=True)
    resource_id: int = Field(
        sa_column=Column(Integer, ForeignKey("resource.id", ondelete="CASCADE"), index=True)
    )
    tag_id: int = Field(
        sa_column=Column(Integer, ForeignKey("resourcetag.id", ondelete="CASCADE"), index=True)
    )
    creation_date: str = ""


class Resource(ResourceBase, table=True):
    __tablename__ = "resource"

    id: int | None = Field(default=None, primary_key=True)
    org_id: int = Field(
        sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True)
    )
    resource_uuid: str = Field(index=True, unique=True)
    created_by_user_id: int | None = Field(
        default=None,
        sa_column=Column(Integer, ForeignKey("user.id", ondelete="SET NULL"), index=True, nullable=True),
    )
    creation_date: str = ""
    update_date: str = ""


class ResourceCreate(ResourceBase):
    tag_uuids: list[str] = Field(default_factory=list)


class ResourceUpdate(SQLModel):
    title: str | None = None
    description: str | None = None
    resource_type: ResourceTypeEnum | None = None
    provider_name: str | None = None
    provider_url: str | None = None
    external_url: str | None = None
    cover_image_url: str | None = None
    estimated_time: int | None = None
    is_featured: bool | None = None
    is_live: bool | None = None
    access_mode: ResourceAccessModeEnum | None = None
    tag_uuids: list[str] | None = None


class ResourceRead(ResourceBase):
    id: int
    org_id: int
    resource_uuid: str
    created_by_user_id: int | None = None
    owner_org_id: int | None = None
    owner_org_uuid: str | None = None
    owner_org_slug: str | None = None
    owner_org_name: str | None = None
    is_shared_from_other_org: bool = False
    creation_date: str
    update_date: str
    tags: list[ResourceTagRead] = Field(default_factory=list)


class ResourceChannelBase(SQLModel):
    name: str
    description: str | None = Field(default=None, sa_column=Column(Text))
    thumbnail_image: str | None = None
    public: bool = True
    shared: bool = False
    is_starred: bool = False
    color: str | None = None


class ResourceChannel(ResourceChannelBase, table=True):
    __tablename__ = "resourcechannel"

    id: int | None = Field(default=None, primary_key=True)
    org_id: int = Field(
        sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True)
    )
    channel_uuid: str = Field(index=True, unique=True)
    creation_date: str = ""
    update_date: str = ""


class ResourceChannelCreate(ResourceChannelBase):
    pass


class ResourceChannelUpdate(SQLModel):
    name: str | None = None
    description: str | None = None
    thumbnail_image: str | None = None
    public: bool | None = None
    shared: bool | None = None
    is_starred: bool | None = None
    color: str | None = None


class ResourceChannelRead(ResourceChannelBase):
    id: int
    org_id: int
    channel_uuid: str
    owner_org_id: int | None = None
    owner_org_uuid: str | None = None
    owner_org_slug: str | None = None
    owner_org_name: str | None = None
    is_shared_from_other_org: bool = False
    creation_date: str
    update_date: str


class ResourceChannelResource(SQLModel, table=True):
    __tablename__ = "resourcechannelresource"
    __table_args__ = (
        UniqueConstraint("channel_id", "resource_id", name="uq_channel_resource"),
    )

    id: int | None = Field(default=None, primary_key=True)
    channel_id: int = Field(
        sa_column=Column(Integer, ForeignKey("resourcechannel.id", ondelete="CASCADE"), index=True)
    )
    resource_id: int = Field(
        sa_column=Column(Integer, ForeignKey("resource.id", ondelete="CASCADE"), index=True)
    )
    sort_order: int = 0
    creation_date: str = ""
    update_date: str = ""


class UserResourceChannelBase(SQLModel):
    name: str
    description: str | None = Field(default=None, sa_column=Column(Text))
    is_default: bool = False
    icon: str | None = None
    color: str | None = None
    icon_color: str | None = None


class UserResourceChannel(UserResourceChannelBase, table=True):
    __tablename__ = "userresourcechannel"

    id: int | None = Field(default=None, primary_key=True)
    user_id: int = Field(
        sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True)
    )
    org_id: int = Field(
        sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True)
    )
    user_channel_uuid: str = Field(index=True, unique=True)
    creation_date: str = ""
    update_date: str = ""


class UserResourceChannelCreate(UserResourceChannelBase):
    pass


class UserResourceChannelRead(UserResourceChannelBase):
    id: int
    user_id: int
    org_id: int
    user_channel_uuid: str
    creation_date: str
    update_date: str


class UserSavedResourceBase(SQLModel):
    notes: str | None = Field(default=None, sa_column=Column(Text))
    outcome_text: str | None = Field(default=None, sa_column=Column(Text))
    outcome_link: str | None = None
    outcome_file: str | None = None
    last_opened_at: str | None = None
    completed_at: str | None = None
    open_count: int = 0


class UserSavedResource(UserSavedResourceBase, table=True):
    __tablename__ = "usersavedresource"
    __table_args__ = (
        UniqueConstraint("user_id", "resource_id", name="uq_user_saved_resource"),
    )

    id: int | None = Field(default=None, primary_key=True)
    user_id: int = Field(
        sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True)
    )
    resource_id: int = Field(
        sa_column=Column(Integer, ForeignKey("resource.id", ondelete="CASCADE"), index=True)
    )
    creation_date: str = ""
    update_date: str = ""


class UserSavedResourceUpdate(SQLModel):
    notes: str | None = None
    outcome_text: str | None = None
    outcome_link: str | None = None
    completed_at: str | None = None
    open_count_increment: int | None = None
    add_to_default_channel: bool = True
    user_channel_uuids: list[str] = []


class UserSavedResourceRead(UserSavedResourceBase):
    id: int
    user_id: int
    resource_id: int
    creation_date: str
    update_date: str


class UserSavedResourceChannel(SQLModel, table=True):
    __tablename__ = "usersavedresourcechannel"
    __table_args__ = (
        UniqueConstraint("saved_resource_id", "user_channel_id", name="uq_saved_resource_channel"),
    )

    id: int | None = Field(default=None, primary_key=True)
    saved_resource_id: int = Field(
        sa_column=Column(Integer, ForeignKey("usersavedresource.id", ondelete="CASCADE"), index=True)
    )
    user_channel_id: int = Field(
        sa_column=Column(Integer, ForeignKey("userresourcechannel.id", ondelete="CASCADE"), index=True)
    )
    creation_date: str = ""


class ResourceCommentBase(SQLModel):
    content: str = Field(sa_column=Column(Text))


class ResourceComment(ResourceCommentBase, table=True):
    __tablename__ = "resourcecomment"

    id: int | None = Field(default=None, primary_key=True)
    resource_id: int = Field(
        sa_column=Column(Integer, ForeignKey("resource.id", ondelete="CASCADE"), index=True)
    )
    author_id: int = Field(
        sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True)
    )
    comment_uuid: str = Field(index=True, unique=True)
    is_locked: bool = Field(default=False, sa_column=Column(Boolean, default=False))
    creation_date: str = ""
    update_date: str = ""


class ResourceCommentCreate(ResourceCommentBase):
    pass


class ResourceCommentUpdate(SQLModel):
    content: str | None = None


class ResourceCommentRead(ResourceCommentBase):
    id: int
    resource_id: int
    author_id: int
    comment_uuid: str
    is_locked: bool
    creation_date: str
    update_date: str


class ResourceCommentReadWithAuthor(ResourceCommentRead):
    author: dict | None = None
